### DISCUSSION TITLE
Axis Camera NTP sync with AWS

### DISCUSSION THREAD

**Discussion by rafiomeon** [AxisCommunications/discussions#90] [VAPIX] [answered]
Is there a way to NTP sync Axis camera time with AWS server(EC2 Instance).

**Comment by mckee3304** [comment, marked as answer]
This worked for me. Update the staticServers param to your NTP address. I'm going to try to iterate through a loop referencing an excel file of IP addresses. I'll update the forum if I am successful.


import requests
import sys

print("Update NTP on Axis Cameras\n")

ip = input("Enter IP Address of Camera: ")
username = input("Enter Username of Camera: ")
password = input("Enter Password of Camera: ")

auth = requests.auth.HTTPDigestAuth(username, password)
body = {
    "apiVersion": "1.1",
    "method": "setNTPClientConfiguration",
    "params": {
        "serversSource": "static", 
        "staticServers": ["time.google.com"]
    }    
}
r = requests.post("http://{}/axis-cgi/ntp.cgi".format(ip), json=body,  auth=auth)

if r.status_code == 200:
    print("\nSuccess")
else: 
    print(r.status_code)
    print(r.text)

**Reply by n04h-zz** [reply]
I'm curious, how did you set up the iteration through the spreadsheet? Did you add additional camera settings in the script?

**Comment by n04h-zz** [comment]
I'm curious, how did you set up the iteration through the spreadsheet? Did you add additional camera settings in the script?

**Reply by mckee3304** [reply]
Here's my final code with a spreadsheet and exporting results to a log file. In the csv file called "devicelist.csv" I had two columns; Device in the first column and IPAddress in the second column. So the csv file looked like this:

Device                     |  IPAddress
Back Yard Camera   |  192.168.1.xxx
Front Yard Camera  |  192.168.1.xxx

I ran this through about 5000 cameras in like an hour. Let me know if you need more of an explanation. 

```
import csv
import os
import requests
import sys
import time
from datetime import datetime, date

# define time
now = datetime.now()

# define script start time
startTime = time.time()
startTimeNormal = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime(time.time())) #converts time from epoch to normal

# define log file name to include filename + time
logSavePath = './'
logFileNameString = str("ScriptLog-")
logFileDate = now.strftime("%m%d%Y")
LogFileDateString = str(logFileDate)
logFileExtension = ".txt"
logFileName =  os.path.join(logSavePath, logFileNameString + LogFileDateString + logFileExtension)

# create log file
logFile = open(logFileName, "a")

# start script 
print("Starting NTP Update on Axis Cameras...\n")
logFile.write("Start Time: " + str(startTimeNormal) + "\n")

# prompt for username & password
username = input("Enter Username of Camera: ")
password = input("Enter Password of Camera: ")
NTPServer = input("Enter new NTP Server IP Address/Hostname: ")

# open list of devices
with open('devicelist.csv', 'r') as csvinput:
    reader = csv.DictReader(csvinput)       
    for row in reader:
        ip = row["IPAddress"]
        auth = requests.auth.HTTPDigestAuth(username, password)
        body = {
            "apiVersion": "1.1",
            "method": "setNTPClientConfiguration",
            "params": {
                "serversSource": "static", 
                "staticServers": [NTPServer]
            }    
        }
        response = os.popen(f"ping {ip} -n 1").read()
        if "Received = 1" in response:   
            r = requests.post("http://{}/axis-cgi/ntp.cgi".format(ip), json=body,  auth=auth)
            if r.status_code == 200:
                device = row["Device"]
                ipAddress = ip
                statusCode = str("Success ")
                # write success to log
                logFile.write(str(statusCode) + "     " + str(ipAddress) + "   " + str(device) + "\n")
            elif r.status_code == 401:
                device = row["Device"]
                ipAddress = ip
                statusCode = str("Authentication Failed ") + str(r.status_code)
                # write authentication failure to log
                logFile.write(str(statusCode) + "     " + str(ipAddress) + "   " + str(device) + "\n")              
            else:
                device = row["Device"]
                ipAddress = ip
                statusCode = str("Failed ") + str(r.status_code)
                # write failed to log
                logFile.write(str(statusCode) + "     " + str(ipAddress) + "   " + str(device) + "\n")             
        else:
            logFile.write(str("Ping Failed") + "     " + str(ipAddress) + "   " + str(device) + "\n")  
            continue     

# calculate time it took to complete script
endTime = time.time()
endTimeNormal = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime(time.time())) #converts time from epoch to normal
executionTime = ((endTime - startTime)/60)
calulatedTime = round(executionTime, 2)

# close log file
logFile.write("End Time: " + str(endTimeNormal) + "\n")
logFile.write("Total Execution Time: " + str(calulatedTime) + " Minutes" + "\n")
logFile.close()

# script complete
print("\n" + "NTP Update Complete!")
```

**Reply by n04h-zz** [reply]
Thank you, I'm going to try this out on a few cameras. I modified that original a bit to include certificate verification and password hashing. I was playing with it working on updating overlays. 

import requests
import sys
import maskpass # hashes the password in the terminal
import certifi #Certificate verification; creates a secure connection to the camera to void MIM issues


print("Update Dynamic Overlay on Axis Camera\n")


name = input("Enter in name of the camera:")
ip = input("Enter IP of Camera:")
username = input("Enter Username of Camera:")
#hashes password
password = maskpass.askpass(prompt="Camera Password:", mask="*")
 
auth = requests.auth.HTTPDigestAuth(username, password)

body = {
    
"apiVersion": "1.0", #api version 
"method": "setText", #method used for each setting and parameters. Get these from the vapix library
"context": "321",
"params": {
"identity": 1,
"text": name + " " + "%F %T %Z",
"position": "bottomLeft",
"textColor": "black"

}
}
#Put the intermediatary and root cert in a .pem format file to use -veryify= option to complete the certificate chain if using a signed certificate on the cameras. File goes into the certifi folder like below. 
r = requests.post("https://{}/axis-cgi/dynamicoverlay/dynamicoverlay.cgi".format(ip), json=body, auth=auth, verify="C:\\PathTo\\Python\\Python310\\Lib\\site-packages\\certifi\\my_trust_store.pem")

if r.status_code == 200:
    print("\nSuccess")
else: 
    print(r.status_code)
    print(r.text)
